src/stream/iter_csv.rs

```diff
@@ -0,0 +1,250 @@
+use csv::{self, Reader, ReaderBuilder};
+use num::Float;
+use std::collections::{HashMap, HashSet};
+use std::marker::PhantomData;
+
+pub struct IterCsv<F: Float + std::str::FromStr, R: std::io::Read> {
+    reader: Reader<R>,
+    headers: csv::StringRecord,
+    y_cols: Option<Target>,
+    data_stream: PhantomData<DataStream<F>>,
+}
+pub enum Target {
+    Name(String),
+    MultipleNames(HashSet<String>),
+}
+
+impl Target {
+    fn contains(&self, name: &str) -> bool {
+        match self {
+            Target::Name(n) => n == name,
+            Target::MultipleNames(names) => names.contains(&name.to_string()),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Data<F: Float + std::str::FromStr> {
+    Scalar(F),
+    String(String),
+}
+
+pub enum DataStream<F: Float + std::str::FromStr> {
+    X(HashMap<String, Data<F>>),
+    XY(HashMap<String, Data<F>>, HashMap<String, Data<F>>),
+}
+
+impl<F: Float + std::str::FromStr> DataStream<F> {
+    pub fn get_x(&self) -> &HashMap<String, Data<F>> {
+        match self {
+            DataStream::X(x) => x,
+            DataStream::XY(x, _) => x,
+        }
+    }
+
+    pub fn get_y(&self) -> Result<&HashMap<String, Data<F>>, &str> {
+        match self {
+            DataStream::X(_) => Err("No y data"),
+            DataStream::XY(_, y) => Ok(y),
+        }
+    }
+}
+
```

Comment:
We should move these structs to another file because it will be used in another iter.